crates/module-system/sov-modules-api/src/state/accessors/scratchpad.rs

```diff
@@ -342,8 +345,27 @@ impl<S: Spec, I: StateProvider<S>> PerBlockCache for TxScratchpad<S, I> {
 
 impl<S: Spec> TxScratchpad<S, StateCheckpoint<S>> {
     /// Change set resulting from transaction execution.
-    pub fn tx_changes(&self) -> TxChangeSet {
-        self.inner.changes()
+    pub fn tx_changes(&self, execution_context: ExecutionContext) -> TxChangeSet {
+        self.inner.changes(execution_context)
+    }
+
+    /// Applies changes to the TxScratchpad to warm up its cache.
+    pub fn apply_change_set(&mut self, changes: TxChangeSet) {
+        let mut reads = changes.reads.unwrap();
+
+        // If a write overrides a read, ignore the read.
+        self.filter_out_writes(&mut reads.user, Namespace::User);
+        self.filter_out_writes(&mut reads.kernel, Namespace::Kernel);
+
+        self.inner.inner.add_read_if_not_present(reads);
```

Comment:
nit: We're paying a bit more than we need to here because we take two passes over the array (one to remove the keys that were written, one to add the new reads) and each pass takes O(capacity) time since these are maps - so the filtering on the first pass doesn't help with the time complexity of the next one